app/services/resource_service.rb

```diff
@@ -12,78 +12,126 @@ def initialize(data_provider:)
+    if @data_provider.user.restricted_role? && resource_class.respond_to?(:visible)
       @params = { visible: false }.merge(@params)
     end
 
     @resource_class = resource_class
     @resource = resource_class.new(@params)
-    @resource.data_provider = data_provider
+    @resource.data_provider = @data_provider
 
     # skip create if record already exists and new record has the same attribute values as the new
     # record and the record is not marked as 'always_recreate'
-    external_resource = find_external_resource
-    old_record = external_resource.try(:external)
-    if external_resource.present? && unchanged_attributes?(old_record) && !always_recreate?(data_provider, resource_class)
-      external_resource.touch
-      return old_record
+    @external_resource = find_external_resource
+    @old_resource = find_old_resource
+    if @external_resource.present? && unchanged_attributes? && !always_recreate?
+      @external_resource.touch
+      return @old_resource
+    end
+
+    resource_or_error_message(update_or_create)
+  end
+
+  # Find old resource
+  #
+  # @return [Object] Instance of @resource_class
+  def find_old_resource
+    if @old_resource_id.present?
+      @resource_class.filtered_for_current_user(@data_provider.user).find_by(id: @old_resource_id)
+    else
+      @external_resource.try(:external)
+    end
+  end
+
+  def update_or_create
+    if @old_resource.present?
+      update_resource
+    else
+      create_new_resource
+    end
+  end
+
+  def update_resource
+    # find all association names to delete for a resource
+    association_names_to_delete = @resource_class
+                                    .reflect_on_all_associations
+                                    .select { |a| a.options[:dependent] == :destroy }.map(&:name)
+
+    # delete all nested resources
+    association_names_to_delete.each do |association_name|
+      associations = @old_resource.send(association_name)
+      associations.destroy_all if !is_singular?(association_name) && associations.any?
+      associations.destroy if is_singular?(association_name) && associations.present?
     end
 
-    create_or_recreate(old_record, external_resource)
-    resource_or_error_message(resource)
+    # update all attributes and recreate nested resources
+    @old_resource.update(@params)
+
+    ExternalReference.create(
+      data_provider: @old_resource.data_provider,
+      external_id: @old_resource.id,
+      external_type: @resource_class,
+      unique_id: @old_resource.unique_id
+    )
+
+    set_default_categories if @resource.respond_to?(:categories)
+
+    @old_resource
   end
 
-  def create_or_recreate(old_record, external_resource)
+  def create_new_resource
     ActiveRecord::Base.transaction do
-      old_record.destroy if old_record.present?
-      external_resource.destroy if external_resource.present?
+      @external_resource.destroy if @external_resource.present?
```

Comment:
wieso muss die external resource noch gelöscht werden? wird in zwei zeilen mit gleichen werten neu erstellt, oder?

Reply:
beim neu Anlegen und beim Aktualisieren  einer Resource müssen eventuelle alte Verweise entfernte werden. Normalerweise sollte das automatisch geschehen über den `dependent: :destroy` Hook. Es kommt aber mittlerweile zu oft vor, dass jemand direkt in der Datenbank Manipulationen vor nimmt und diese Daten nicht mit bereinigt.